community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/BatchingMultipleIndexPopulator.java

```diff
@@ -0,0 +1,418 @@
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.kernel.impl.api.index;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.RejectedExecutionHandler;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Supplier;
+
+import org.neo4j.function.Predicates;
+import org.neo4j.helpers.Exceptions;
+import org.neo4j.kernel.api.exceptions.index.IndexEntryConflictException;
+import org.neo4j.kernel.api.exceptions.index.IndexPopulationFailedKernelException;
+import org.neo4j.kernel.api.index.IndexConfiguration;
+import org.neo4j.kernel.api.index.IndexDescriptor;
+import org.neo4j.kernel.api.index.IndexPopulator;
+import org.neo4j.kernel.api.index.NodePropertyUpdate;
+import org.neo4j.kernel.api.index.SchemaIndexProvider;
+import org.neo4j.logging.LogProvider;
+import org.neo4j.storageengine.api.schema.PopulationProgress;
+import org.neo4j.unsafe.impl.internal.dragons.FeatureToggles;
+
+import static java.util.stream.Collectors.joining;
+import static org.neo4j.helpers.NamedThreadFactory.daemon;
+
+/**
+ * A {@link MultipleIndexPopulator} that gathers all incoming updates from the {@link IndexStoreView} in batches of
+ * size {@link #BATCH_SIZE} and then flushes each batch from different thread using {@link ExecutorService executor}.
+ * <p>
+ * It is possible for concurrent updates from transactions to arrive while index population is in progress. Such
+ * updates are inserted in the queue. When store scan notices that queue size has reached {@link #QUEUE_THRESHOLD} than
+ * it drains all batched updates and waits for all submitted to the executor tasks to complete and flushes updates from
+ * the queue using {@link MultipleIndexUpdater}. If queue size never reaches {@link #QUEUE_THRESHOLD} than all queued
+ * concurrent updates are flushed after the store scan in {@link #flipAfterPopulation()}.
+ * <p>
+ * Inner {@link ExecutorService executor} is shut down after the store scan completes.
+ */
+public class BatchingMultipleIndexPopulator extends MultipleIndexPopulator
+{
+    public static final String QUEUE_THRESHOLD_NAME = "queue_threshold";
```

Comment:
This particular field is used in `MultipleIndexPopulationStressIT` which lives in a different package. Made following three fields package-private.